File: clients/java/lib/src/main/java/health/varta/Varta.java
package health.varta;

import health.varta.errno.ErrnoClassifier;
import health.varta.transport.BeatTransport;
import health.varta.transport.UdpTransport;
import health.varta.transport.UdsTransport;
import health.varta.vlp.Codec;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.file.Path;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.zip.CRC32C;

/**
 * Varta agent — emit 32-byte VLP heartbeats to a local observer.
 *
 * <p>Use the static factories to connect ({@link #connect(Path)},
 * {@link #connectUdp(InetSocketAddress)},
 * {@link #connectSecureUdp(InetSocketAddress, Key)},
 * {@link #connectSecureUdpWithMaster(InetSocketAddress, Key)}), then call
 * {@link #beat(Status, int)} on a fixed cadence (typically every 500 ms).</p>
 *
 * <p>{@code beat()} is non-blocking and never throws. A return of
 * {@link BeatOutcome.Dropped} indicates either kernel backpressure or
 * observer-absent — both are recoverable conditions the application can
 * log or ignore. {@link BeatOutcome.Failed} signals an unexpected error.</p>
 *
 * <p>The class is thread-safe: a single intrinsic monitor serialises all
 * {@code beat()} calls. Concurrent emit from many threads is supported
 * but the call rate is bounded by the cost of one
 * {@code System.nanoTime} + one non-blocking {@code send(2)}.</p>
 */
public final class Varta implements AutoCloseable {
    public static final int FRAME_BYTES = Codec.FRAME_BYTES;
    public static final long NONCE_TERMINAL = Codec.NONCE_TERMINAL;
    public static final long NONCE_MIN = Codec.NONCE_MIN;
    private static final AtomicBoolean NONCE_WRAP_WARNED = new AtomicBoolean(false);

    private final Object lock = new Object();
    private final BeatTransport transport;
    private final byte[] scratch = new byte[FRAME_BYTES];
    private final ByteBuffer scratchBuf = ByteBuffer.wrap(scratch).order(ByteOrder.LITTLE_ENDIAN);
    private final CRC32C crc = new CRC32C();
    private final long startNanos = System.nanoTime();

    private int connectPid;
    private long nonce = NONCE_MIN;
    private long lastTimestamp = 0L;
    private long clockRegressions = 0L;
    private long forkRecoveries = 0L;
    private int reconnectAfter = 0;
    private int consecutiveDropped = 0;
    private boolean closed = false;

    private Varta(BeatTransport transport) {
        this.transport = transport;
        this.connectPid = currentPid();
    }

    /** Connect over UDS (AF_UNIX SOCK_DGRAM). Requires a UDS provider on the classpath. */
    public static Varta connect(Path socketPath) {
        Objects.requireNonNull(socketPath, "socketPath");
        try {
            return new Varta(UdsTransport.create(socketPath));
        } catch (IOException e) {
            throw new IllegalStateException("Varta.connect: " + e.getMessage(), e);
        }
    }

    /** Connect over plaintext UDP. Recovery commands are refused by the observer. */
    public static Varta connectUdp(InetSocketAddress addr) {
        Objects.requireNonNull(addr, "addr");
        try {
            return new Varta(UdpTransport.create(addr));
        } catch (IOException e) {
            throw new IllegalStateException("Varta.connectUdp: " + e.getMessage(), e);
        }
    }

    /**
     * Connect over ChaCha20-Poly1305 AEAD UDP using a 32-byte shared key.
     * The wire frame is 60 bytes; AAD is empty. Recovery is refused.
     * <p>Implementation lives in {@code health.varta.transport.SecureUdpTransport}
     * — wired in Phase 3.</p>
     */
    public static Varta connectSecureUdp(InetSocketAddress addr, Key key) {
        Objects.requireNonNull(addr, "addr");
        Objects.requireNonNull(key, "key");
        try {
            return new Varta(health.varta.transport.SecureUdpTransport.createShared(addr, key));
        } catch (IOException e) {
            throw new IllegalStateException("Varta.connectSecureUdp: " + e.getMessage(), e);
        }
    }

    /**
     * Connect over ChaCha20-Poly1305 AEAD UDP using a 32-byte master key.
     * The wire frame is 64 bytes; AAD = {@code u32_LE(agentPid)}. Recovery is refused.
     */
    public static Varta connectSecureUdpWithMaster(InetSocketAddress addr, Key masterKey) {
        Objects.requireNonNull(addr, "addr");
        Objects.requireNonNull(masterKey, "masterKey");
        try {
            return new Varta(health.varta.transport.SecureUdpTransport.createMaster(addr, masterKey));
        } catch (IOException e) {
            throw new IllegalStateException("Varta.connectSecureUdpWithMaster: " + e.getMessage(), e);
        }
    }

    /** Emit one beat with payload = 0. Convenience for the common case. */
    public BeatOutcome beat(Status status) {
        return beat(status, 0);
    }

    /** Emit one beat with the supplied 32-bit payload. Non-blocking; never throws. */
    public BeatOutcome beat(Status status, int payload) {
        if (status == null) {
            return BeatOutcome.failed(new BeatError(0, "InvalidInput"));
        }
        synchronized (lock) {
            if (closed) {
                return BeatOutcome.failed(new BeatError(0, "Varta agent is closed"));
            }

            // 1) Fork-safety: PID changed → reconnect, reset session state.
            int currentPid = currentPid();
            if (currentPid != connectPid) {
                try {
                    transport.reconnect();
                    connectPid = currentPid;
                    nonce = NONCE_MIN;
                    lastTimestamp = 0L;
                    consecutiveDropped = 0;
                    forkRecoveries = saturatingInc(forkRecoveries);
                } catch (IOException e) {
                    // Fork-recovery reconnect failed: the fork invalidated the
                    // old socket and a new one could not be established, so the
                    // beat path is broken. This is a terminal error, NOT a
                    // transient drop — return Failed unconditionally, matching
                    // the Rust reference (BeatOutcome::Failed(BeatError::from_io))
                    // and every other client (Go "ReconnectFailed", Python/Node
                    // from_oserror/fromNodeError, .NET). Routing it through
                    // ErrnoClassifier would misclassify e.g. "connection refused"
                    // as Dropped(NO_OBSERVER), telling the caller the beat path
                    // is still operational when fork recovery has actually
                    // failed — so it would keep retrying a dead beat loop instead
                    // of surfacing the hard failure.
                    return BeatOutcome.failed(new BeatError(0, "ReconnectFailed"));
                }
            }

            // 2) Monotonic timestamp + regression clamp. Compute the candidate
            // high-water WITHOUT committing it; commit-on-success advances it
            // only when send accepts the datagram.
            long nowNs = System.nanoTime() - startNanos;
            if (nowNs < lastTimestamp) {
                clockRegressions = saturatingInc(clockRegressions);
                nowNs = lastTimestamp;
            }
            long candidateTimestamp = nowNs;

            // 3) Sentinel reservation: u64::MAX is reserved on the wire.
            long wireTimestamp = nowNs;
            if (wireTimestamp == 0xFFFF_FFFF_FFFF_FFFFL) {
                wireTimestamp = 0xFFFF_FFFF_FFFF_FFFEL;
            }

            // 4) Nonce sequencing — compute the wire nonce and the value the
            // counter advances to WITHOUT committing either until send succeeds
            // (commit-on-success). A Dropped or Failed attempt leaves the same
            // candidate for the next beat, so no invisible nonce/timestamp is
            // burned on the wire. Mirrors crates/varta-client/src/client.rs
            // (next_regular_nonce / commit_sent_frame).
            long candidateNonce = nonce;
            long nextNonce;
            boolean wrappedNonce = false;
            if (candidateNonce == NONCE_TERMINAL) {
                // NONCE_TERMINAL is reserved for panic-hook Critical frames.
                // The regular stream wraps to wire nonce 0 and only commits
                // to nonce 1 after the send succeeds.
                candidateNonce = 0L;
                nextNonce = NONCE_MIN;
                wrappedNonce = true;
            } else {
                nextNonce = candidateNonce + 1;
            }

            // 5) Encode + send. On a Dropped outcome, mirror the cross-client
            // contract: after `reconnectAfter` consecutive drops, reconnect and
            // retry once (recovery from an observer restart). The counter resets
            // on any Sent or Failed outcome. Nonce/timestamp commit only for a
            // frame the kernel actually accepted.
            Codec.encodeInto(scratchBuf, status, currentPid, wireTimestamp, candidateNonce, payload, crc);
            BeatOutcome outcome = sendScratch();
            if (outcome instanceof BeatOutcome.Sent) {
                commitSentFrame(nextNonce, candidateTimestamp, wrappedNonce);
                consecutiveDropped = 0;
                return outcome;
            }
            if (outcome instanceof BeatOutcome.Dropped) {
                consecutiveDropped = saturatingIncInt(consecutiveDropped);
                if (reconnectAfter > 0 && consecutiveDropped >= reconnectAfter) {
                    consecutiveDropped = 0;
                    boolean reconnected;
                    try {
                        transport.reconnect();
                        reconnected = true;
                    } catch (IOException e) {
                        reconnected = false;
                    }
                    if (reconnected) {
                        BeatOutcome retry = sendScratch();
                        if (retry instanceof BeatOutcome.Sent) {
                            commitSentFrame(nextNonce, candidateTimestamp, wrappedNonce);
                        }
                        return retry;
                    }
                }
                return outcome;
            }
            // Failed: reset like a Sent so a transient error does not arm a
            // spurious reconnect on the next drop. Nonce/timestamp stay
            // uncommitted for the next beat.
            consecutiveDropped = 0;
            return outcome;
        }
    }

    /**
     * Advance the committed nonce/timestamp after the kernel accepted the
     * datagram (commit-on-success).
     */
    private void commitSentFrame(long nextNonce, long timestamp, boolean wrappedNonce) {
        nonce = nextNonce;
        lastTimestamp = timestamp;
        if (wrappedNonce) {
            warnNonceWrapping();
        }
    }

    private static void warnNonceWrapping() {
        if (NONCE_WRAP_WARNED.compareAndSet(false, true)) {
            System.err.println("[varta] nonce exhausted; wrapping to 0");
        }
    }

    /** Send the already-encoded scratch buffer once and classify the result. */
    private BeatOutcome sendScratch() {
        scratchBuf.rewind();
        try {
            int written = transport.send(scratchBuf);
            if (written == 0) {
                return BeatOutcome.dropped(DropReason.KERNEL_QUEUE_FULL);
            }
            if (written != FRAME_BYTES) {
                return BeatOutcome.failed(new BeatError(0, "WriteZero"));
            }
            return BeatOutcome.sent();
        } catch (IOException e) {
            return ErrnoClassifier.classify(e);
        }
    }

    /** Reopen the underlying transport. Throws {@link IllegalStateException} on I/O failure. */
    public void reconnect() {
        synchronized (lock) {
            try {
                transport.reconnect();
                connectPid = currentPid();
                consecutiveDropped = 0;
            } catch (IOException e) {
                throw new IllegalStateException("Varta.reconnect: " + e.getMessage(), e);
            }
        }
    }

    /**
     * Auto-reconnect after {@code n} consecutive {@link BeatOutcome.Dropped}
     * outcomes — recovery from an observer restart. The internal counter
     * increments on each Dropped beat and resets to zero on any Sent or Failed
     * outcome; once it reaches {@code n} the next beat reconnects and retries
     * once. {@code 0} (default) disables auto-reconnect. Mirrors Rust's
     * {@code Varta::set_reconnect_after}.
     */
    public void setReconnectAfter(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be >= 0");
        synchronized (lock) {
            this.reconnectAfter = n;
            this.consecutiveDropped = 0;
        }
    }

    /** Saturating counter of detected clock regressions. */
    public long clockRegressions() {
        synchronized (lock) {
            return clockRegressions;
        }
    }

    /** Saturating counter of fork-driven reconnects. */
    public long forkRecoveries() {
        synchronized (lock) {
            return forkRecoveries;
        }
    }

    @Override
    public void close() {
        synchronized (lock) {
            if (closed) return;
            closed = true;
            transport.close();
        }
    }

    private static long saturatingInc(long v) {
        return v == Long.MAX_VALUE ? v : v + 1;
    }

    private static int saturatingIncInt(int v) {
        return v == Integer.MAX_VALUE ? v : v + 1;
    }

    private static int currentPid() {
        return (int) ProcessHandle.current().pid();
    }

    // Test hooks — package-private, no public stability contract.

    static Varta __forTest(BeatTransport transport) {
        return new Varta(transport);
    }

    void __setConnectPidForTest(int pid) {
        synchronized (lock) { this.connectPid = pid; }
    }

    void __setNonceForTest(long n) {
        synchronized (lock) { this.nonce = n; }
    }

    long __getNonceForTest() {
        synchronized (lock) { return nonce; }
    }
}

File: clients/dotnet/src/Varta.Client/Varta.cs
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Net.Sockets;
using Varta.Internal.Errno;
using Varta.Internal.Transport;
using Varta.Internal.Vlp;
using Varta.Internal.VlpSecure;

namespace Varta;

/// <summary>
/// Varta agent — emits VLP v0.2 health beats to a local observer over a
/// configured transport. Construct via one of the <c>Connect*</c>
/// factories, then call <see cref="Beat"/> on whatever cadence the
/// application chooses (typically every 500 ms – 5 s).
/// </summary>
/// <remarks>
/// <para>
/// Thread-safe via a single internal mutex; concurrent <see cref="Beat"/>
/// callers serialise. The mutex is uncontended in the common case of one
/// thread per agent.
/// </para>
/// <para>
/// Fork-safety: <see cref="Beat"/> compares the live
/// <see cref="Environment.ProcessId"/> against the PID snapshot taken at
/// <see cref="Connect"/> time. On mismatch (i.e. running in a forked
/// child) the transport is rebuilt — for secure UDP that re-reads the
/// session salt from OS entropy. The recovery count is exposed via
/// <see cref="ForkRecoveries"/>.
/// </para>
/// </remarks>
[SuppressMessage("Naming", "CA1724:Type names should not match namespaces",
    Justification = "The class name 'Varta' is the agreed public surface — matches the package id Varta.Client and the namespace convention used by all peer clients (Rust crates/varta-client::Varta, Go varta.Varta, Python varta.Varta).")]
public sealed class Varta : IDisposable
{
    /// <summary>Wire size of a VLP v0.2 base frame.</summary>
    public const int FrameBytes = Frame.Bytes;

    /// <summary>
    /// Sentinel nonce signalling a terminal Critical beat. Only legal on
    /// the wire paired with <see cref="Status.Critical"/>.
    /// </summary>
    public const ulong NonceTerminal = Frame.NonceTerminal;

    private readonly IBeatTransport _transport;
    private readonly object _lock = new();
    private readonly byte[] _scratch = new byte[FrameBytes];
    private readonly long _startTimestamp = Stopwatch.GetTimestamp();

    private int _connectPid;
    private ulong _nonce = 1;
    private ulong _lastTimestamp;
    private ulong _clockRegressions;
    private ulong _forkRecoveries;
    private uint _consecutiveDropped;
    private uint _reconnectAfter;
    private bool _disposed;

    private Varta(IBeatTransport transport)
    {
        _transport = transport;
        _connectPid = Environment.ProcessId;
    }

    /// <summary>
    /// Connect to a Varta observer over a Unix Domain Socket
    /// (<c>AF_UNIX</c> + <c>SOCK_DGRAM</c>). Linux and macOS only —
    /// throws <see cref="PlatformNotSupportedException"/> on Windows
    /// (use <see cref="ConnectUdp"/> instead).
    /// </summary>
    public static Varta Connect(string socketPath)
    {
        ArgumentNullException.ThrowIfNull(socketPath);
        return new Varta(new UdsTransport(socketPath));
    }

    /// <summary>
    /// Connect via plaintext UDP. Dev / loopback use only — production
    /// deployments should prefer <see cref="Connect"/> (UDS,
    /// kernel-attested) or <see cref="ConnectSecureUdp"/>
    /// (AEAD-authenticated).
    /// </summary>
    public static Varta ConnectUdp(string host, int port)
    {
        ArgumentNullException.ThrowIfNull(host);
        return new Varta(new UdpTransport(host, port));
    }

    /// <summary>
    /// Connect via ChaCha20-Poly1305 AEAD UDP using a 32-byte pre-shared
    /// key. Produces 60-byte wire frames.
    /// </summary>
    public static Varta ConnectSecureUdp(string host, int port, ReadOnlySpan<byte> key32)
    {
        ArgumentNullException.ThrowIfNull(host);
        return new Varta(new SecureUdpTransport(SecureUdpMode.Shared, host, port, key32));
    }

    /// <summary>
    /// Connect via ChaCha20-Poly1305 AEAD UDP using a 32-byte master
    /// key; the per-agent key is HKDF-derived from the master key and
    /// the current PID. Produces 64-byte wire frames; agent_pid is bound
    /// as AAD.
    /// </summary>
    public static Varta ConnectSecureUdpWithMaster(string host, int port, ReadOnlySpan<byte> masterKey32)
    {
        ArgumentNullException.ThrowIfNull(host);
        return new Varta(new SecureUdpTransport(SecureUdpMode.Master, host, port, masterKey32));
    }

    /// <summary>
    /// Emit one beat. The call is non-blocking; a full kernel queue
    /// surfaces as <see cref="BeatOutcome.Dropped"/> with
    /// <see cref="DropReason.KernelQueueFull"/> — never an exception.
    /// </summary>
    public BeatOutcome Beat(Status status, uint payload = 0)
    {
        lock (_lock)
        {
            ObjectDisposedException.ThrowIf(_disposed, this);
            if (!IsAgentStatus(status))
            {
                _consecutiveDropped = 0;
                return BeatOutcome.Failed(new BeatError(0, "InvalidInput"));
            }

            // Per-emission PID read — never cache.
            int currentPid = Environment.ProcessId;
            if (currentPid != _connectPid)
            {
                try
                {
                    _transport.Reconnect();
                }
                catch (SocketException ex)
                {
                    // Beat() must never throw (the documented contract). A failed
                    // fork-recovery reconnect is surfaced as Failed — matching the
                    // Rust reference (client.rs: `Err(e) => BeatOutcome::Failed`)
                    // and the Go/Python/Node clients. _connectPid is left
                    // unchanged so the next beat retries the reconnect.
                    return BeatOutcome.Failed(
                        new BeatError(ex.NativeErrorCode, ex.SocketErrorCode.ToString()));
                }
                _connectPid = currentPid;
                _nonce = 1;
                _lastTimestamp = 0;
                _forkRecoveries = SaturatingIncrement(_forkRecoveries);
            }

            // Compute the nonce and timestamp CANDIDATES without committing
            // them. They advance only when Send accepts the datagram
            // (commit-on-success): a Dropped or Failed attempt leaves the same
            // candidate available for the next beat, so no invisible
            // nonce/timestamp is burned on the wire. Mirrors
            // crates/varta-client/src/client.rs (next_regular_nonce /
            // commit_sent_frame).
            //
            // Monotonic ns since this Varta was constructed. Stopwatch.GetElapsedTime
            // is unaffected by NTP adjustments. TimeSpan.Ticks = 100 ns.
            ulong nowNs = SaturatingNanosFromTicks(Stopwatch.GetElapsedTime(_startTimestamp).Ticks);
            if (nowNs < _lastTimestamp)
            {
                _clockRegressions = SaturatingIncrement(_clockRegressions);
                nowNs = _lastTimestamp;
            }
            ulong candidateTimestamp = nowNs;

            ulong candidateNonce = _nonce;
            // The committed _nonce becomes this on success; wraps past the
            // reserved terminal sentinel back to 0.
            ulong nextNonce = _nonce < NonceTerminal - 1 ? _nonce + 1 : 0;

            Codec.EncodeInto(_scratch, status, (uint)currentPid, candidateTimestamp, candidateNonce, payload);

            return TrySendWithReconnectAfter(allowRetry: true, nextNonce, candidateTimestamp);
        }
    }

    private BeatOutcome TrySendWithReconnectAfter(bool allowRetry, ulong nextNonce, ulong candidateTimestamp)
    {
        try
        {
            int sent = _transport.Send(_scratch);
            if (sent != FrameBytes)
            {
                _consecutiveDropped = 0;
                return BeatOutcome.Failed(new BeatError(0, "WriteZero"));
            }
            CommitSentFrame(nextNonce, candidateTimestamp);
            _consecutiveDropped = 0;
            return BeatOutcome.Sent();
        }
        catch (SocketException ex)
        {
            var outcome = ErrnoClassifier.Classify(ex);
            if (!outcome.IsDropped)
            {
                // Failed: reset like a Sent so a transient error does not
                // arm a spurious reconnect on the next drop. The nonce and
                // timestamp are left uncommitted for the next beat.
                _consecutiveDropped = 0;
                return outcome;
            }
            _consecutiveDropped = SaturatingIncrementU32(_consecutiveDropped);
            if (allowRetry && _reconnectAfter > 0 && _consecutiveDropped >= _reconnectAfter)
            {
                _consecutiveDropped = 0;
                try
                {
                    _transport.Reconnect();
                    _connectPid = Environment.ProcessId;
                }
                catch
                {
                    return outcome;
                }
                return TrySendWithReconnectAfter(allowRetry: false, nextNonce, candidateTimestamp);
            }
            return outcome;
        }
    }

    // Advance the committed nonce/timestamp only after the kernel accepted the
    // datagram (commit-on-success).
    private void CommitSentFrame(ulong nextNonce, ulong timestamp)
    {
        _nonce = nextNonce;
        _lastTimestamp = timestamp;
    }

    /// <summary>
    /// Force a transport rebuild and refresh the connect-time PID
    /// snapshot. For secure UDP, also re-reads the session salt from OS
    /// entropy. Useful after a known fork(2) or to recover from a
    /// half-open peer.
    /// </summary>
    public void Reconnect()
    {
        lock (_lock)
        {
            _transport.Reconnect();
            _connectPid = Environment.ProcessId;
            _consecutiveDropped = 0;
        }
    }

    /// <summary>
    /// Enable auto-reconnect after <paramref name="n"/> consecutive
    /// dropped beats. Set to 0 to disable. Counter resets on any Sent
    /// outcome.
    /// </summary>
    public void SetReconnectAfter(uint n)
    {
        lock (_lock)
        {
            _reconnectAfter = n;
            _consecutiveDropped = 0;
        }
    }

    /// <summary>
    /// Saturating count of fork(2) auto-recoveries detected by
    /// <see cref="Beat"/>. Publish this as a Prometheus gauge under your
    /// chosen metric name (suggested: <c>varta_client_fork_recoveries_total</c>).
    /// </summary>
    public ulong ForkRecoveries
    {
        get { lock (_lock) { return _forkRecoveries; } }
    }

    /// <summary>
    /// Saturating count of detected platform-clock regressions
    /// (<see cref="Stopwatch"/> regressions are extraordinarily rare but
    /// possible on virtualised hosts). Publish as a Prometheus gauge
    /// under e.g. <c>varta_client_clock_regression_total</c>.
    /// </summary>
    public ulong ClockRegressions
    {
        get { lock (_lock) { return _clockRegressions; } }
    }

    public void Dispose()
    {
        lock (_lock)
        {
            if (_disposed) return;
            _disposed = true;
            _transport.Dispose();
        }
    }

    // ---- Test hooks (InternalsVisibleTo Varta.Client.Tests) ----

    internal void SetConnectPidForTest(int pid)
    {
        lock (_lock) { _connectPid = pid; }
    }

    internal static Varta FromTransportForTest(IBeatTransport transport) => new(transport);

    internal ulong NonceForTest
    {
        get { lock (_lock) { return _nonce; } }
        set { lock (_lock) { _nonce = value; } }
    }

    // ---- Helpers ----

    private static ulong SaturatingIncrement(ulong v) => v == ulong.MaxValue ? v : v + 1;
    private static uint SaturatingIncrementU32(uint v) => v == uint.MaxValue ? v : v + 1;
    private static bool IsAgentStatus(Status status) =>
        status == Status.Ok || status == Status.Degraded || status == Status.Critical;

    // Convert 100-ns Stopwatch ticks to nanoseconds, saturating instead of
    // overflowing. `ticks * 100` is signed-long arithmetic; after a
    // multi-century single-handle uptime it would overflow `long` into a
    // negative value whose `(ulong)` cast lands near the reserved BadTimestamp
    // sentinel (u64::MAX), which the observer drops. Mirrors the panic path
    // (SignalHandler.BuildCriticalFrame) and the Rust client's
    // `self.start.elapsed().as_nanos().min(u64::MAX as u128) as u64`.
    // internal for the regression test (InternalsVisibleTo Varta.Client.Tests).
    internal static ulong SaturatingNanosFromTicks(long ticks) =>
        ticks > long.MaxValue / 100 ? (ulong)long.MaxValue : (ulong)(ticks * 100L);
}
